File: server/logger.js
import { AsyncLocalStorage } from 'node:async_hooks';
import { config } from './config.js';

export const requestStore = new AsyncLocalStorage();

const LEVELS = { debug: 10, info: 20, warn: 30, error: 40 };
const configuredLevel = LEVELS[config.logLevel] || (config.isProduction ? LEVELS.info : LEVELS.debug);

function write(level, message, meta = {}) {
  if (LEVELS[level] < configuredLevel) return;
  const ctx = requestStore.getStore() || {};
  const entry = {
    ts: new Date().toISOString(),
    level,
    service: config.service.name,
    env: config.env,
    requestId: ctx.requestId,
    msg: message,
    ...meta
  };
  const line = JSON.stringify(entry);
  if (level === 'error') console.error(line);
  else if (level === 'warn') console.warn(line);
  else console.log(line);
}

export const logger = {
  debug: (msg, meta) => write('debug', msg, meta),
  info: (msg, meta) => write('info', msg, meta),
  warn: (msg, meta) => write('warn', msg, meta),
  error: (msg, meta) => write('error', msg, meta),
  child(baseMeta = {}) {
    return {
      debug: (msg, meta) => write('debug', msg, { ...baseMeta, ...meta }),
      info: (msg, meta) => write('info', msg, { ...baseMeta, ...meta }),
      warn: (msg, meta) => write('warn', msg, { ...baseMeta, ...meta }),
      error: (msg, meta) => write('error', msg, { ...baseMeta, ...meta })
    };
  }
};

File: server/engine/logger.py
import json
import logging
from datetime import datetime, timezone

from .config import config


class JsonFormatter(logging.Formatter):
    def format(self, record):
        log_entry = {
            "ts": datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"),
            "level": record.levelname.lower(),
            "service": config.Service.NAME,
            "env": config.ENV,
            "msg": record.getMessage()
        }
        if hasattr(record, "meta"):
            log_entry.update(record.meta)
        return json.dumps(log_entry)

logger = logging.getLogger("radly")
logger.setLevel(logging.INFO if config.IS_PRODUCTION else logging.DEBUG)
handler = logging.StreamHandler()
handler.setFormatter(JsonFormatter())
logger.addHandler(handler)

class LoggerProxy:
    @staticmethod
    def _log(level, msg, meta=None):
        extra = {"meta": meta} if meta else {}
        if level == "debug": logger.debug(msg, extra=extra)
        elif level == "info": logger.info(msg, extra=extra)
        elif level == "warn": logger.warning(msg, extra=extra)
        elif level == "error": logger.error(msg, extra=extra)

    @classmethod
    def debug(cls, msg, meta=None): cls._log("debug", msg, meta)
    
    @classmethod
    def info(cls, msg, meta=None): cls._log("info", msg, meta)
    
    @classmethod
    def warn(cls, msg, meta=None): cls._log("warn", msg, meta)
    
    @classmethod
    def error(cls, msg, meta=None): cls._log("error", msg, meta)

radly_logger = LoggerProxy()
